server/extension/extension.go

```diff
@@ -277,81 +405,113 @@ func (m *Manager) registerExtensions(r *mux.Router, extConfigs *ExtensionConfigs
 	return nil
 }
 
+// authorize will enforce rbac rules are satified for the given RequestResources.
+// The following validations are executed:
+//   - enforce the subject has permission to read application/project provided
+//     in HeaderArgoCDApplicationName and HeaderArgoCDProjectName.
+//   - enforce the subject has permission to invoke the extension identified by
+//     extName.
+//   - enforce that all resources provided in the HeaderArgoCDResourceGVKName belong
+//     to the application identified by HeaderArgoCDApplicationName.
+//
+// If all validations are satified it will return the Application resource
+func (m *Manager) authorize(ctx context.Context, rr *RequestResources, extName string) (*v1alpha1.Application, error) {
+	if m.rbac == nil {
+		return nil, fmt.Errorf("rbac enforcer not set in extension manager")
+	}
+	appRBACName := security.AppRBACName(rr.ApplicationNamespace, rr.ProjectName, rr.ApplicationNamespace, rr.ApplicationName)
+	if err := m.rbac.EnforceErr(ctx.Value("claims"), rbacpolicy.ResourceApplications, rbacpolicy.ActionGet, appRBACName); err != nil {
+		return nil, fmt.Errorf("application authorization error: %s", err)
+	}
+
+	if err := m.rbac.EnforceErr(ctx.Value("claims"), rbacpolicy.ResourceExtensions, rbacpolicy.ActionInvoke, extName); err != nil {
+		return nil, fmt.Errorf("unauthorized to invoke extension %q: %s", extName, err)
+	}
+
+	// just retrieve the app after checking if subject has access to it
+	app, err := m.application.Get(rr.ApplicationNamespace, rr.ApplicationName)
+	if err != nil {
+		return nil, fmt.Errorf("error getting application: %s", err)
+	}
+	if app == nil {
+		return nil, fmt.Errorf("invalid Application provided in the %q header", HeaderArgoCDApplicationName)
+	}
+
+	if app.Spec.GetProject() != rr.ProjectName {
+		return nil, fmt.Errorf("project mismatch provided in the %q header", HeaderArgoCDProjectName)
```

Comment:
A failure in this check should report the same error message as a failed RBAC check would. Otherwise we risk leaking the existence of an Application which is in a project which the user is not allowed to access.

Reply:
I intentionally want this to be a different message so we can track back in the code when we see it in the logs. Having the same error message makes troubleshooting much harder.